### PR TITLE
[Wip] Smarty3 - branch for testing

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
[Wip] Smarty3 - branch for testing

This adds code that makes it possible to

1) enabled Smarty3 as an extension
2) toggle back & forth between Smarty3  & Smarty2 with the define `define('CIVICRM_SMARTY3_AUTOLOAD_PATH', \path\to\civicrm\ext\smarty3\vendor\autoload.php`);`

Most of what this code adds is already put up as separate PRs as it is mergeable now

What the code adds is 
1) a smarty3 extension that really ONLY functions to allow the composer package to be present but condtionally loaded
2) https://github.com/civicrm/civicrm-core/pull/27585 `SmartyCompatibility.php`this provides a function that allows v3 Smarty functions to run in v2. This could be added to core ASAP allowing us to migrate functions like `clear_all_templates()` to `clearAllTemplates()`
3) Some minor fixes that could be pulled out & merged like not having unneed variables passed by reference (`crmScope`),  , https://github.com/civicrm/civicrm-core/pull/27581 https://github.com/civicrm/civicrm-core/pull/27579 https://github.com/civicrm/civicrm-core/pull/27583
4) *the removal of our fetch function override* - https://github.com/civicrm/civicrm-core/pull/27588 this was our attempt to impose security by raising security whenever `fetch` received a `:string`. We have another method that does this `parseOneOffStringThroughSmarty` & perhaps we could use deprecation notices to find places that should be moved to that to allow us to remove this function. The signture changes
5) replace use of Smarty.get to get url variables with assigns - these cause notices in smartyv3 (maybe not again i v4?) https://github.com/civicrm/civicrm-core/pull/27582
6) function smarty_modifier_smarty($string, $class) {` (this is not implemented in v3 the same as v2) is a hack to push out default escaping I think - I did that a while back

**After this is installed the most broken things are**

1) Our good old help helpers
image
2) pages with backticks https://github.com/civicrm/civicrm-core/pull/27547#issuecomment-1730950671

Before
----------------------------------------

After
----------------------------------------

Technical Details
----------------------------------------
@larssandergreen with this you can get close to testing out your changes. I'm just trying to figure out if I previously solved 2

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Core/SmartyCompatibility.php [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ */
+
+/**
+ * Fix for bug CRM-392. Not sure if this is the best fix or it will impact
+ * other similar PEAR packages. doubt it
+ */
+if (!class_exists('Smarty')) {
+  if (defined('CIVICRM_SMARTY3')) {
+    global $civicrm_root;
+    require_once $civicrm_root . '/ext/smarty3/vendor/autoload.php';
```
Comment (larssandergreen) [inline CRM/Core/SmartyCompatibility.php:25]: This file doesn't exist. Should it? Or is it supposed to `ext/smarty3/vendor/smarty/smarty/lib/AutoLoader.php`?

Comment (larssandergreen) [inline CRM/Core/SmartyCompatibility.php:25]: Also, shouldn't it be like
`CRM_Utils_File::addTrailingSlash($civicrm_root) . 'ext...`

**Comment by larssandergreen** [comment]
Seeing a bunch of notices like this:
```
Notice: Indirect modification of overloaded property CRM_Core_Smarty::$template_dir has no effect in _afform_html_civix_civicrm_config()
```
Smarty upgrade docs say:
== $cache_dir, $compile_dir, $config_dir, $template_dir access ==
The mentioned properties can't be accessed directly any longer. You must use
corresponding getter/setters like addConfigDir(), setConfigDir(), getConfigDir()

So I guess we need an upgrade to civix for this and something to keep existing extensions working?

**Comment by eileenmcnaughton** [comment]
@larssandergreen this should address that https://github.com/civicrm/civicrm-core/pull/27579

**Comment by larssandergreen** [comment]
@eileenmcnaughton We can do that for core extensions, but it's not just that extension, it's any extension that's not updated with civix recently (on my local, that's civirules, civivisualize, angular profiles, log viewer, contribution recur, twilio, iats and mosaico as well).  So I guess then we get into a bigger question of how to handle upgrading to Smarty 3 without causing issues with existing extensions.

For example, contribution pages using IATS payment processing are now broken.

**Comment by eileenmcnaughton** [comment]
@larssandergreen yeah - I think the first goal would be to get all the pages working with Smarty3 in core....

Probably worth doing a `civix upgrade` PR for any that you hit though? I assume they will pick up the mixin?

**Comment by larssandergreen** [comment]
@eileenmcnaughton Sure, I can do some of those later too.

**Comment by eileenmcnaughton** [comment]
@larssandergreen this search string `& CRM_Core_Smarty::singleton` seems to work

**Comment by JoeMurray** [comment]
@Edzelopez could you assign some of these upgrades to our staff, potentially Mary or Lauren or Monish. Lars is it okay if we start with civirules, contribution recur and iats ?

**Comment by eileenmcnaughton** [comment]
@JoeMurray I did put up an IATS pr

**Comment by larssandergreen** [comment]
@JoeMurray @Edzelopez That would be great. Note that the list of extensions above are just the ones I happen to have had on my local. Any extension with civix version of 23.01 or lower should be upgraded.

**Comment by eileenmcnaughton** [comment]
I just updated this so the define you need (if you git pull which you don't have to) is now `CIVICRM_SMARTY3_AUTOLOAD_PATH` - will update PR

**Comment by JoeMurray** [comment]
@Edzelopez let's also go through our extensions starting with line item edit and the grant ones and other financial ones.

**Comment by monishdeb** [comment]
For now I have submitted the MRs/PRs for respective extension's civix upgrade below: 

Civix upgrade PRs: 
1. CiviRule - https://lab.civicrm.org/extensions/civirules/-/merge_requests/221 
1. ActivityType ACL - https://lab.civicrm.org/extensions/activitytypeacl/-/merge_requests/29 
1. Contribution Recur - https://github.com/adixon/ca.civicrm.contributionrecur/pull/19 
1. Civisualize - https://lab.civicrm.org/extensions/civisualize/-/merge_requests/118 
1. Log Viewer - https://github.com/adixon/ca.civicrm.logviewer/pull/18 
1. Grant Applications - https://lab.civicrm.org/extensions/grantapplications/-/merge_requests/196
1. Angular Profile - https://lab.civicrm.org/extensions/angularprofiles/-/merge_requests/1 
1. CiviMoodle - https://lab.civicrm.org/extensions/civimoodle/-/merge_requests/20 
1. Mandrill Email - https://lab.civicrm.org/extensions/mte/-/merge_requests/121
1. Outbound Domain Enforcement - https://lab.civicrm.org/extensions/ode/-/merge_requests/29

**Comment by eileenmcnaughton** [comment]
thanks @monishdeb - what do you still use angular profiles for ? I guess it's still part of civivolunteer?

**Comment by eileenmcnaughton** [comment]
Note the stuff in this branch is otherwise merged

**Pull request closed without merging**